nameserver/dns_test.go

```diff
@@ -1,39 +1,68 @@
-	. "github.com/weaveworks/weave/common"
+
+	"github.com/weaveworks/weave/net/address"
+	"github.com/weaveworks/weave/router"
 )
 
-// Check that we can prune an answer
-func TestPrune(t *testing.T) {
-	EnableDebugLogging(testing.Verbose())
-	Log.Infoln("TestPrune starting")
-
-	questionMsg := new(dns.Msg)
-	questionMsg.SetQuestion("name", dns.TypeA)
-	questionMsg.RecursionDesired = true
-	question := &questionMsg.Question[0]
-	records := []ZoneRecord{
-		Record{"name", net.ParseIP("10.0.1.1"), 0, 0, 0},
-		Record{"name", net.ParseIP("10.0.1.2"), 0, 0, 0},
-		Record{"name", net.ParseIP("10.0.1.3"), 0, 0, 0},
-		Record{"name", net.ParseIP("10.0.1.4"), 0, 0, 0},
+func TestLargeUDPResonse(t *testing.T) {
+	//common.SetLogLevel("debug")
+
+	peername, err := router.PeerNameFromString("00:00:00:02:00:00")
+	require.Nil(t, err)
+	nameserver := New(peername, nil, nil, "")
+	dnsserver, err := NewDNSServer(nameserver, "weave.local.", 0, 30, 5*time.Second)
+	require.Nil(t, err)
+	udpPort := dnsserver.servers[0].PacketConn.LocalAddr().(*net.UDPAddr).Port
+	tcpPort := dnsserver.servers[1].Listener.Addr().(*net.TCPAddr).Port
+	go dnsserver.ActivateAndServe()
+	defer dnsserver.Stop()
+
+	// Add 100 mappings to nameserver
+	addrs := []address.Address{}
+	for i := address.Address(0); i < 100; i++ {
+		addrs = append(addrs, i)
+		nameserver.AddEntry("foo.weave.local.", "", peername, i)
 	}
 
-	reply := makeAddressReply(questionMsg, question, records, DefaultLocalTTL)
-	reply.Answer[0].Header().Ttl = DefaultLocalTTL
+	doRequest := func(client *dns.Client, request *dns.Msg, port int) *dns.Msg {
+		request.SetQuestion("foo.weave.local.", dns.TypeA)
+		response, _, err := client.Exchange(request, fmt.Sprintf("127.0.0.1:%d", port))
+		require.Nil(t, err)
+		return response
+	}
 
-	pruned := pruneAnswers(reply.Answer, 1)
-	require.Equal(t, 1, len(pruned), "wrong number of answers")
+	// do a udp query, ensure we get a truncated response
+	{
+		udpClient := dns.Client{Net: "udp", UDPSize: minUDPSize}
+		response := doRequest(&udpClient, &dns.Msg{}, udpPort)
+		require.Nil(t, err)
+		require.True(t, response.MsgHdr.Truncated)
+	}
 
-	pruned = pruneAnswers(reply.Answer, 2)
-	require.Equal(t, 2, len(pruned), "wrong number of answers")
+	// do a udp query with big size, ensure we don't get a truncated response
+	{
+		udpClient := dns.Client{Net: "udp", UDPSize: 65535}
+		request := &dns.Msg{}
+		request.SetEdns0(65535, false)
+		response := doRequest(&udpClient, request, udpPort)
+		require.False(t, response.MsgHdr.Truncated)
+		require.Equal(t, len(response.Answer), 100)
+	}
 
-	pruned = pruneAnswers(reply.Answer, 0)
-	require.Equal(t, len(records), len(pruned), "wrong number of answers")
+	// do a tcp query, ensure we don't get a truncated response
+	{
+		tcpClient := dns.Client{Net: "tcp"}
+		response := doRequest(&tcpClient, &dns.Msg{}, tcpPort)
+		require.False(t, response.MsgHdr.Truncated)
+		require.Equal(t, len(response.Answer), 100)
+	}
 }
```

Comment:
We have tests for unsupported query types in the integration tests.

Re malformed queries, not FQDNs etc - I don't see the point of adding these tests.  (1) the parsing is not down to us, but meikg/dns and (2) we know there will fail.